components/suggest/src/yelp.rs

```diff
@@ -142,299 +153,260 @@ impl SuggestDao<'_> {
-            let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
-            if self.is_modifier(&candidate, pre_modifier_type)? {
-                pre_modifier = Some(candidate);
-                break;
-            }
-        }
+        // This checks if keyword is a substring of the query.
+        if let Some(keyword_lowercase) = self.conn.try_query_one::<String, _>(
+            "SELECT keyword
+             FROM yelp_subjects
+             WHERE :query BETWEEN keyword AND keyword || ' ' || x'FFFF'
+             ORDER BY LENGTH(keyword) ASC, keyword ASC
+             LIMIT 1",
+            named_params! {
+                ":query": query_string.to_lowercase(),
+            },
+            true,
+        )? {
+            // Preserve the query as the user typed it including its case.
+            let keyword = &query_string[0..keyword_lowercase.len()];
+            let count = keyword.split_whitespace().count();
+            return Ok(Some((
+                keyword.to_string(),
+                true,
+                query_words.get(count..).unwrap_or_default(),
+            )));
+        };
 
-        let mut post_modifier = None;
-        for n in (1..=MAX_MODIFIER_WORDS_NUMBER).rev() {
-            let mut candidate_chunks = words.rchunks(n);
-            let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
-            if self.is_modifier(&candidate, post_modifier_type)? {
-                post_modifier = Some(candidate);
-                break;
-            }
+        if query_string.len() < SUBJECT_PREFIX_MATCH_THRESHOLD {
+            return Ok(None);
         }
 
-        let mut without_modifiers = query;
-        if let Some(ref modifier) = pre_modifier {
-            without_modifiers = &without_modifiers[modifier.len()..];
-        }
-        if let Some(ref modifier) = post_modifier {
-            without_modifiers = &without_modifiers[..without_modifiers.len() - modifier.len()];
-        }
+        // Oppositely, this checks if the query is a substring of keyword.
+        if let Some(keyword_lowercase) = self.conn.try_query_one::<String, _>(
+            "SELECT keyword
+             FROM yelp_subjects
+             WHERE keyword BETWEEN :query AND :query || x'FFFF'
+             ORDER BY LENGTH(keyword) ASC, keyword ASC
+             LIMIT 1",
+            named_params! {
+                ":query": query_string.to_lowercase(),
+            },
+            true,
+        )? {
+            // Preserve the query as the user typed it including its case.
+            let keyword = format!(
+                "{}{}",
+                query_string,
+                &keyword_lowercase[query_string.len()..]
+            );
+
+            let count = std::cmp::min(query_words.len(), keyword.split_whitespace().count());
+            return Ok(Some((
+                keyword,
+                false,
+                query_words.get(count..).unwrap_or_default(),
+            )));
```

Comment:
```suggestion
            return Ok(keyword_lowercase
                .get(query_string.len()..)
                .map(|keyword_rest| {
                    query_string.push_str(keyword_rest);
                    let count =
                        std::cmp::min(query_words.len(), query_string.split_whitespace().count());
                    (
                        query_string,
                        false,
                        query_words.get(count..).unwrap_or_default(),
                    )
                }));
```

Similar here. Also, `format!` is probably slower than just appending the rest of the keyword to the query string. With my suggested change, you'll need to declare `query_string` as `mut`.

Reply:
I see, thanks!